libcaf_core/caf/action.hpp

```diff
@@ -132,17 +132,22 @@ inline bool operator!=(const action& lhs, const action& rhs) noexcept {
 namespace caf::detail {
 
 template <class F, bool IsSingleShot>
-struct default_action_impl : detail::atomic_ref_counted, action::impl {
-  std::atomic<action::state> state_;
-  F f_;
-
+class default_action_impl : public detail::atomic_ref_counted,
+                            public action::impl {
+public:
   default_action_impl(F fn)
     : state_(action::state::scheduled), f_(std::move(fn)) {
     // nop
   }
 
+  ~default_action_impl() {
+    if (state_.load() == action::state::scheduled)
+      f_.~F();
+  }
+
   void dispose() override {
-    state_ = action::state::disposed;
+    state_.store(action::state::disposed);
+    f_.~F();
```

Comment:
Should try to set `state_` to `disposed` atomically and only call the destructor when transitioning from another state.

```
    state_.store(action::state::disposed);
    f_.~F();
```